src/main/java/games/strategy/triplea/ui/MapPanel.java

```diff
@@ -823,42 +817,31 @@ void showMouseCursor() {
     return uiContext.getMapData().getWarningImage();
   }
 
-  private static final class BackgroundDrawer implements Runnable {
-    private MapPanel mapPanel;
-
-    BackgroundDrawer(final MapPanel panel) {
-      mapPanel = panel;
-    }
+  private final class BackgroundDrawer implements Runnable {
+    private volatile boolean running = true;
 
     void stop() {
-      // the thread will eventually wake up and notice we are done
-      mapPanel = null;
+      running = false;
     }
 
     @Override
     public void run() {
-      while (mapPanel != null) {
-        final BlockingQueue<Tile> undrawnTiles;
-        final MapPanel panel = mapPanel;
-        undrawnTiles = panel.getUndrawnTiles();
-        final Tile tile;
+      while (running) {
         try {
-          tile = undrawnTiles.poll(2000, TimeUnit.MILLISECONDS);
+          final Tile tile = undrawnTiles.poll(2, TimeUnit.SECONDS);
+          if (tile != null) {
+            final GameData data = MapPanel.this.getData();
```

Comment:
There's something about this that's bugging me.  It almost seems as if the tuple `(tile, gameData)` should have been sent through the queue rather than just `tile` itself.  There's nothing stopping a tile from being added to the `undrawnTile` queue with one `gameData` in place, and then someone calling `setGameData()` before that tile is dequeued and drawn, so that the value returned on this line will be different than what existed when the tile was added to the queue.  I don't know if that mismatch would cause a problem.

Also, access to `MapPanel#gameData` is not synchronized, but it is being read/written from different threads.  That could also cause this thread to use the wrong `GameData` value until memory becomes consistent again.  (Sending it through the queue would make this issue moot.)

These problems existed before this PR, so I'm not suggesting that you fix them, as you've already addressed the root cause of the NPE, but it might be something to consider.  What's your opinion, @RoiEXLab?

Reply:
Well I haven't really looked into the functionality, but I agree that the access of this Thread in general needs to be more synchronised in general, also because of the issue I mentioned above, so the loop is instantly being aborted, once the game is being stopped.
In any case something for a follow-up PR.